### PR TITLE
Fixing consumer to timeout on pubsub disconnection or network issues

### PR DESCRIPTION
**Pull request by pranavcode**
- Adding timeout (30 secs deadline) for couple of scenarios where go-metrics gets stuck as mentioned in (4) and (5).

1. Go-metrics app is connected to PubSub, PubSub dies, Go-metrics detect the disconnection and retries connecting forever, connects when PubSub is available [InPlace]
2. Go-metrics cannot connect to PubSub, Go-metrics detect the disconnection and retries connecting forever, connects when PubSub is available [InPlace]
3. Go-metrics is connected to PubSub, Go-metrics disconnects, PubSub disconnects, Go-metrics tries to reconnect and connects when PubSub is available [InPlace]
4. Go-metrics is connected to pubsub, PubSub is not listening any more, Go-metrics app is stuck (exit and restart go-metrics) [Solved By Timeout]
5. Go-metrics tries connecting to PubSub, PubSub is not receiving any connections, Go-metrics app is stuck (exit and restart go-metrics) [Solved By Timeout]

- Adding additional consumer check for subscription
- Google PubSub's go module has new method updates recently as;
  - `Message`'s `Done(bool)` is now divided in two different acknowledgement methods `Ack()` / `Nack()`
  - `Subscription`'s `Pull(context)` is now `Receive(context, callback)`
- Additionally;
  - Adding glide go package manager.

Connect to https://github.com/replaygaming/go-metrics/issues/10

### CODE REVIEW

File: consumer_test.go [hunk 1]
```diff
@@ -75,6 +75,18 @@ func TestRemove(t *testing.T) {
...
 }
 
+func TestAlive(t *testing.T) {
+	pubsubClient, _ := pubsub.NewClient(context.Background(), "emulator-project-id")
+	subscription := pubsubClient.Subscription("test-subscription")
+
+	c := NewConsumer("test-topic", "test-subscription")
+	c.Remove()
+
+	if !c.Alive() {
```
Comment (Draiken) [inline consumer_test.go]: Isn't this `!` extra here?

File: pubsub_consumer.go [hunk 2]
```diff
@@ -32,9 +33,13 @@ func (m *googlePubSubMessage) Data() []byte {
 	return m.OriginalMessage.Data
 }
 
-// Delegate to pubsub message's Done
+// Delegate to pubsub message's Ack/Nack
 func (m *googlePubSubMessage) Done(ack bool) {
-	m.OriginalMessage.Done(ack)
+	if ack == true {
```
Comment (Draiken) [inline pubsub_consumer.go]: I don't recall what this `ack` stands for. You're sure it's the ack/nack change?

File: pubsub_consumer.go [hunk 3]
```diff
@@ -108,8 +117,11 @@ func ensureTopic(pubsubClient *pubsub.Client, topicName string) *pubsub.Topic {
 func ensureSubscription(pubsubClient *pubsub.Client, topic *pubsub.Topic, subscriptionName string) *pubsub.Subscription {
 	var subscription *pubsub.Subscription
 	subscription = pubsubClient.Subscription(subscriptionName)
-	subscriptionExists, _ := subscription.Exists(context.Background())
+	subscriptionExists, err := subscription.Exists(context.Background())
 
+	if err != nil {
+		log.Fatalf("PubSub subscription does not exist: %v", err)
```
Comment (Draiken) [inline pubsub_consumer.go]: Wouldn't the error here be something more like: could not check if subscription exists?

File: pubsub_consumer.go [hunk 4]
```diff
@@ -91,8 +97,11 @@ func NewConsumer(topicName string, subscriptionName string) Consumer {
 func ensureTopic(pubsubClient *pubsub.Client, topicName string) *pubsub.Topic {
 	var topic *pubsub.Topic
 	topic = pubsubClient.Topic(topicName)
-	topicExists, _ := topic.Exists(context.Background())
+	topicExists, err := topic.Exists(context.Background())
 
+	if err != nil {
+		log.Fatalf("PubSub topic does not exist: %v", err)
```
Comment (Draiken) [inline pubsub_consumer.go]: Wouldn't the error here be something more like: could not check if topic exists?

We already try to create the topic afterwards, so the `err` would only be present on other cases

File: pubsub_consumer.go [hunk 5]
```diff
@@ -123,31 +135,35 @@ func ensureSubscription(pubsubClient *pubsub.Client, topic *pubsub.Topic, subscr
 
 // Creates a channel that pulls messages from the subscription
 func (consumer *googlePubSubConsumer) Consume() (chan Message, error) {
+	var mu sync.Mutex
```
Comment (Draiken) [inline pubsub_consumer.go]: Why do we need this mutex?
We're dispatching messages to a channel, which is a goroutine. Not sure why we would need to lock this.

File: pubsub_consumer.go [hunk 6]
```diff
@@ -123,31 +135,35 @@ func ensureSubscription(pubsubClient *pubsub.Client, topic *pubsub.Topic, subscr
...
 
-			wrappedMsg := &googlePubSubMessage{OriginalMessage: msg}
+				wrappedMsg := &googlePubSubMessage{OriginalMessage: msg}
+				channel <- wrappedMsg
+				msg.Ack()
+			})
 
-			channel <- wrappedMsg
+		if err != nil {
+			cancel()
```
Comment (Draiken) [inline pubsub_consumer.go]: Don't think it makes sense to cancel here. The receive method [only returns if][1] there is an unrecoverable error, otherwise it blocks.

[1]: https://godoc.org/cloud.google.com/go/pubsub#Subscription.Receive

File: pubsub_consumer.go [hunk 7]
```diff
@@ -123,31 +135,35 @@ func ensureSubscription(pubsubClient *pubsub.Client, topic *pubsub.Topic, subscr
...
-			}
+		err := consumer.Subscription.Receive(cctx,
+			func(ctx context.Context, msg *pubsub.Message) {
+				mu.Lock()
+				defer mu.Unlock()
 
-			wrappedMsg := &googlePubSubMessage{OriginalMessage: msg}
+				wrappedMsg := &googlePubSubMessage{OriginalMessage: msg}
+				channel <- wrappedMsg
+				msg.Ack()
```
Comment (Draiken) [inline pubsub_consumer.go]: This means that we send the message to the consumer, and acknowledge the message regardless if the message was processed or not.

I think the consumer of the messages should be responsible for acknowledging them. There are cases where we want to acknowledge even if we error out, but there are others where it doesn't make sense. If this behavior is on the `consumer` package, the libraries that use it will have no choice.

I actually think the message has a `Done` method that does `ack`, so we can't really call `ack` here.

File: pubsub_consumer.go (hunk 3 above)
Comment (pranavcode) [inline pubsub_consumer.go]: Yeah, I will reword it to say so for both subscription and topic.

File: pubsub_consumer.go (hunk 2 above)
Comment (pranavcode) [inline pubsub_consumer.go]: Yes. This ack is to make sure the message is notified to be successfully processed and consumed, to be removed from pubsub. The new library has made it use specific methods for positive and negative acknowledgment.

File: consumer_test.go (hunk 1 above)
Comment (pranavcode) [inline consumer_test.go]: It is added for negative assertion.

Comment (Draiken) [inline consumer_test.go]: Then the message is wrong: "Expected consumer not to be alive" :stuck_out_tongue:

Comment (pranavcode) [inline consumer_test.go]: I think we are on a different page here.

Given the subscription is removed from the topic and my aliveness test is based on the existence of subscription, the test fails if `Alive` succeeds and vice versa. So, if the negative assertion on `Alive` passes, we have a failing test, which would need to tell, what is expected. Which is why the error says `Expected consumer to not be alive`, as there is not subscription in existence for the given topic.

File: pubsub_consumer.go (hunk 5 above)
Comment (pranavcode) [inline pubsub_consumer.go]: I understand your concern. will be removing the mutex, it's counter-intuitive to block go-routine, maybe if this was handled in some other case, we would have implemented mutex, but need a quick clarification, how is the message synchronization handled when the messages are received from pubsub and sent on channel to amplitude? afaiu, all these systems are communicating on network and there might be issues with async messages.

File: pubsub_consumer.go (hunk 7 above)
Comment (pranavcode) [inline pubsub_consumer.go]: Message in this context is pubsub.Message, which no longer has Done method associated to it, [that method now has been converted to two different methods - Ack and Nack](https://github.com/GoogleCloudPlatform/google-cloud-go/commit/1f6cb61c9e1ca4af0385fd194041afee7ba7af95).

File: consumer_test.go (hunk 1 above)
Comment (Draiken) [inline consumer_test.go]: You're doing this: If not alive then print "expected not to be alive"
See the problem there?

Either you're expecting it to BE alive, so the message is wrong, or you're expecting it to NOT to be alive, then the conditional is wrong

File: pubsub_consumer.go [hunk 8]
```diff
@@ -135,35 +136,21 @@ func ensureSubscription(pubsubClient *pubsub.Client, topic *pubsub.Topic, subscr
...
-				mu.Lock()
-				defer mu.Unlock()
-
 				wrappedMsg := &googlePubSubMessage{OriginalMessage: msg}
 				channel <- wrappedMsg
-				msg.Ack()
 			})
 
 		if err != nil {
-			cancel()
```
Comment (Draiken) [inline pubsub_consumer.go]: I think I was wrong on this... apparently the `cancel()` method is used to cleanup resources, so I think what you had was correct.

File: pubsub_consumer.go [hunk 9]
```diff
@@ -96,14 +95,15 @@ func NewConsumer(topicName string, subscriptionName string) Consumer {
 // Finds or creates a topic
 func ensureTopic(pubsubClient *pubsub.Client, topicName string) *pubsub.Topic {
 	var topic *pubsub.Topic
+	ctx, _ := context.WithTimeout(context.Background(), 5 * time.Second)
```
Comment (Draiken) [inline pubsub_consumer.go]: 30 seconds should be good to avoid false-negatives

File: pubsub_consumer.go (hunk 8 above)
Comment (Draiken) [inline pubsub_consumer.go]: Check the example on their documentation, I think it shows exactly when they call this `cancel`.

File: pubsub_consumer.go (hunk 9 above)
Comment (pranavcode) [inline pubsub_consumer.go]: I have already made it have 30 seconds timeout, will be in next commit. :+1:

File: pubsub_consumer.go (hunk 8 above)
Comment (pranavcode) [inline pubsub_consumer.go]: Going through the documentation again, I have come to realize, we did it unnecessarily actually. And you made a valid point about it before. Receive will internally call `cancel()` on `context` and release the resources if required. Here is the documentation reference from Google PubSub's Receive method - `The context passed to f will be canceled when ctx is Done or there is a fatal service error`. Code reference [here](https://github.com/GoogleCloudPlatform/google-cloud-go/blob/master/pubsub/subscription.go#L264). That is what I have learnt, what do you say?

Comment (Draiken) [inline pubsub_consumer.go]: Sounds reasonable. Also we're probably not even going to care, since after the receive method returns, we're going to exit anyway...

Comment (pranavcode) [inline pubsub_consumer.go]: :+1:

File: .gitignore [hunk 10]
```diff
@@ -0,0 +1,3 @@
+vendor
+src
```
Comment (Draiken) [inline .gitignore]: Why `src`?

File: pubsub_consumer.go [hunk 11]
```diff
@@ -90,11 +95,15 @@ func NewConsumer(topicName string, subscriptionName string) Consumer {
 // Finds or creates a topic
 func ensureTopic(pubsubClient *pubsub.Client, topicName string) *pubsub.Topic {
 	var topic *pubsub.Topic
+	ctx, _ := context.WithTimeout(context.Background(), 30 * time.Second)
```
Comment (Draiken) [inline pubsub_consumer.go]: Since we use this timeout on several places with the same number (and I don't see why they'd be different on each place) we can use a constant for it

Comment (pranavcode) [inline pubsub_consumer.go]: Valid point. Code smell: magic numbers.

File: .gitignore (hunk 10 above)
Comment (pranavcode) [inline .gitignore]: I was working with GOPATH pointing to the same location as the project, not `~/go`, so `go build` would need imports to be under `src`. So, I soft linked `vendor` to `src`. But, this is not required, very system state specific.

**Comment by Draiken** [comment]
:shipit:
